### PR TITLE
Wrap exceptions in WebServiceException

### PR DESCRIPTION
**Pull request by bgrainger**

### CODE REVIEW

File: src/Faithlife.WebRequests/WebServiceRequestBase.cs [hunk 1]
```diff
@@ -329,6 +336,24 @@ private void SetCookie(HttpResponseHeaders headers, Uri requestUri)
...
+		private static WebServiceException WrapRequestException(HttpRequestMessage webRequest, Exception exception)
+		{
+			(exception as WebException).Response?.Dispose();
+			return new WebServiceException(message: "Error building request.", requestMethod: webRequest.Method.Method, requestUri: webRequest.RequestUri, innerException: exception);
+		}
+
+		private static bool ShouldWrapException(Exception exception)
+		{
+			// InvalidDataException can be thrown if GZip stream header returned from server is invalid.
+			// ObjectDisposedException can be thrown by Mono.
```
Comment (ejball) [inline src/Faithlife.WebRequests/WebServiceRequestBase.cs]: You're not catching this. Delete the comment?

File: src/Faithlife.WebRequests/WebServiceRequestBase.cs [hunk 2]
```diff
@@ -329,6 +336,24 @@ private void SetCookie(HttpResponseHeaders headers, Uri requestUri)
...
 
+		private static WebServiceException WrapRequestException(HttpRequestMessage webRequest, Exception exception)
+		{
+			(exception as WebException).Response?.Dispose();
+			return new WebServiceException(message: "Error building request.", requestMethod: webRequest.Method.Method, requestUri: webRequest.RequestUri, innerException: exception);
+		}
+
+		private static bool ShouldWrapException(Exception exception)
+		{
+			// InvalidDataException can be thrown if GZip stream header returned from server is invalid.
```
Comment (ejball) [inline src/Faithlife.WebRequests/WebServiceRequestBase.cs:341]: I wonder if this is true for `HttpClient`, but better safe than sorry?

File: src/Faithlife.WebRequests/WebServiceRequestBase.cs [hunk 3]
```diff
@@ -329,6 +336,24 @@ private void SetCookie(HttpResponseHeaders headers, Uri requestUri)
...
+			// ObjectDisposedException can be thrown by Mono.
+			return
+				exception is WebException ||
```
Comment (ejball) [inline src/Faithlife.WebRequests/WebServiceRequestBase.cs]: I don't think `HttpClient` is supposed to throw `WebException`, but maybe it still does in exceptional circumstances.

Comment (bgrainger) [reply src/Faithlife.WebRequests/WebServiceRequestBase.cs]: I could live with not catching it, to start with. I don't think it should be thrown. I've only ever seen it wrapped inside another exception (`HttpRequestException` or `IOException`).

Comment (bgrainger) [reply src/Faithlife.WebRequests/WebServiceRequestBase.cs]: Then I'll delete the `WebException.Response.Dispose` code, too.

File: src/Faithlife.WebRequests/WebServiceRequestBase.cs (hunk 2 above)
Comment (bgrainger) [reply src/Faithlife.WebRequests/WebServiceRequestBase.cs:341]: I haven't been able to repro the failure (with `HttpWebRequest` or `HttpClient`) using a simple HTTP server that's designed to return invalid data. I might just keep it, though.

**Review by ejball — APPROVED** [review]

**Pull request merged**